very minimalistic support for colors (warnings and errors)

c-cube opened the pull request:
The point of this PR is to poll maintainers and users, do NOT merge (yet) :-)

The idea is simple: print warnings in yellow and errors in red, as other compilers (gcc, clang for instance) do. I didn't spend a lot of time on this (no tests, a problem with `Format` which indents weirdly because of escape codes). I could use `ANSITerminal` but I'm not sure about adding dependencies to the compiler.

Drup commented:
You should really use [Format.tag](http://caml.inria.fr/pub/docs/manual-ocaml/libref/Format.html#TYPEtag) for this. It would simplify the enabling/disabling and solve the weird indentation.

Respecting `GCC_COLORS` (iirc) might be a good idea.

gasche commented:
My personal opinion is that while I don't personally care about colors (Emacs and Merlin do that for me), I have heard enough demand from enough people to believe that integrating it in the compiler is worthwhile. I find it rather silly to say that a compiler is "massively more usable" when it adds colors (not saying that's what @c-cube claims, but I have heard this argument around the Clang debates), but if enough people hold that irrational belief it becomes an important aspect of usability to please them -- and also they might be right, who knows?

Cc: @ppedrot which asked me privately about doing such things in the toplevel (but I think there `utop` may be the ultimate answer).

c-cube commented:
@gasche I think many people use OCaml without emacs, those days, including me. Colors do not have a "massive" effects, but they do help spot where important messages lie.
I'll follow @Drup's advice on Format tags, although it requires a bit more refactoring to ensure all messages are built using Format.

dbuenzli commented:
One of the problem with tags is that they are not compositional (though it may not be a problem in that case, it is if you try to provide styling as a library). You can also simply use `"@<0>"` to solve the weird identation, this is the way [this API](http://erratique.ch/software/bos/doc/Bos.Fmt.html#styled) is [implemented](https://github.com/dbuenzli/bos/blob/3da12e9d36c3d1df9cd72f2ffca17132ee5301e9/src/bos_fmt.ml#L131-L164).

objmagic commented:
I like this idea, personally. Colored warnings and errors are definitely helpful, especially for beginners.

c-cube commented:
The code is better now (using Format tags). It prints warnings in magenta, locations in bold, errors in red.

c-cube commented:
I should probably add some support into ocamlbuild (with a tag?)

gasche commented:
Yes, I was thinking of that yesterday. If you don't mind, an ocamlbuild tag would be nice -- it's rather easy to add by imitating what is already present, eg `-opaque` for example. You should also adapt the `OCAMLPARAM` parsing, I suppose.

damiendoligez commented:
Three questions:
1. Any thoughts (from anyone) about whether colors should be enabled by default?
2. Does Emacs deal with them, or do they have to be disabled when compiling under Emacs?
3. Should the new code be in a new file?

Also, you definitely need to add this option to the `OCAMLPARAM` code.

Drup commented:
for 2. there is always the possibility to check terminfo ... which should be done anyway. Emacs users can enable [ansicolor](http://www.emacswiki.org/emacs/AnsiColor).

c-cube commented:
There seems to be some `Terminfo` handling in the compiler, but I have no idea how that works. Also, a lot of printing is made into buffers rather than directly into stdout/stderr.

whitequark commented:
"Massively more usable" is a silly claim, but "massively more readable" isn't one. For that matter, even OCaml agrees and uses some formatting for the toplevel.

gasche commented on driver/compenv.ml:
```diff
@@ -209,6 +209,17 @@ let read_OCAMLPARAM ppf position =
+      (* color output *)
+      | "color" ->
+          begin try
+            let b = bool_of_string v in
```
There already is a standardized interface for boolean parameters, which is to use `0`/`1` (see [Compenv.setter](https://github.com/ocaml/ocaml/blob/trunk/driver/compenv.ml#L127-L138)) rather than `bool_of_string`. You should feel free to generalize `Compenv.setter` to support calling an arbitrary function (as an auxiliary function).

gasche commented on driver/main.ml:
```diff
@@ -125,6 +125,7 @@ module Options = Main_args.Make_bytecomp_options (struct
   let _w = (Warnings.parse_options false)
   let _warn_error = (Warnings.parse_options true)
   let _warn_help = Warnings.help_warnings
+  let _color = Location.enable_color
```
I am unconvinced that this is the right interface for this code. It would be more standard to have a `Cflags.color : bool ref`. Doing this would force you to have some logic to initialize color-printing (a `Location.initialize_color` function doing much of what this one does today) on the first location to print; but that would not increase the code complexity much as you already have this only-on-the-first-call logic anyway.

Also something global outside `Location` scales better to a future where other aspects of the compiler output could be colored (`-help`, `-warn-help` for example).

gasche commented on parsing/location.ml:
```diff
@@ -250,6 +251,110 @@ let get_pos_info pos =
 ;;
 
+(* basic color handling *)
+module Color = struct
```
This should go to `misc` (I think `misc/utils.ml` or `misc/terminfo.ml` would be reasonable; creating a new `misc/ansi_colors.ml` appears to be a nice choice, but there is a small risk of module conflicts among people linking compiler-libs), except for the `styles` stuff which makes sense in Location. `mark_{open,close}_tag` could either stay there or be parametrized over a `tag:string -> style list option` mapping.

gasche replied:
Thinking twice: the `styles` stuff could remain inside `Color` if we consider that it should centralize all color styles in the compiler distribution (and happens to only contain location/error/warning stuff for now). Having the information that is currently in `default_styles` centralized in the same place may make it easier for people to customize default colors in the future. So regarding the location of the `styles` stuff, I would accept whatever @c-cube feels is best.

gasche commented:
I think the current codebase is reasonable and could be mergeable after having been made a bit less `Location`-centric. (Once we open the colored box, I am sure that plenty of people will have ideas for even better color output, so it makes sense to put it in a central place from the start.)

Re. default setting: I would enable it by default on `TERM` values that are not empty or `dumb`. (I just checked: it seems that the toplevel started by Emacs modes is set with `dumb`, so that would disable it there by default.). There seem to be `terminfo` settings to retrive color support information (eg. maximal number of colors), but I would wait for a PR of someone willing to implement this.

On the other hand, checking how other projects handle this question (eg. clang?) may suffice to have a good/consensual answer.

whitequark commented:
@gasche Usually, you check the `TERM` and also look at the result of `isatty(fileno(stdout))`. This is what clang does.

c-cube commented:
I'm working on @gasche 's recommendations. I moved `Color` to `Misc.Color` (for fear of module conflicts) with some one-shot initialization code. I put a flag `Clflags.color` (but it's not a boolean, it's a ternary value `On | Off | Auto`). The value `Auto` enables colors depending on `Sys.getenv "TERM"` (`isatty` is only available in `Unix`, which is not currently linked, right?).

gasche commented on driver/main_args.ml:
```diff
@@ -368,6 +368,14 @@ let mk_warn_help f =
+let mk_no_color f =
+  "-no-color", Arg.Unit f,
+  " <bool> Disable ANSI color codes for warning/error messages"
+
```
It would be nice if there was a way to explicitly ask on the command line the "auto" mode -- you had a good idea in separating these three options. You could have `-color-auto` as a third flag, or maybe accept `-color {on,off,auto}` (or `{always,never,auto}`, these are the option names in git and grep). This would mean that the `OCAMLPARAM` handling code would have to be changed again...

gasche commented on ocamlbuild/ocaml_specific.ml:
```diff
@@ -638,6 +638,8 @@ let () =
   pflag ["ocaml"; "native"; "compile"] "inline"
     (fun param -> S [A "-inline"; A param]);
+  pflag ["ocaml"; "compile"] "color"
+    (fun param -> S [A "-color"; A param]);
```
This may need to be updated when you are settled on a command-line usage.

c-cube commented:
Indeed, forgot about the ocamlbuild tag (which should become several tags: `color`, `no_color` and `color_auto`, I guess). The past has shown it was good to provide an explicit option for each possible value of a parameter (including the default one), so I agree we need `-color-auto` and the corresponding OCAMLPARAM flag (maybe a 'a').

gasche commented:
I'm not sure why the testsuite fails, but it may be the case that some tests in the testsuite do not run with `TERM=dumb` while they should (I have had issues with this in the past because the location-printing code in the toplevel depends on `TERM`, and some tests run a toplevel: e2b662980fe1baba29c96a10195eed9a0a71859c ).

c-cube commented:
The errors were because the output was slightly changed by this branch; the last commit should fix this.

c-cube commented:
yay!

gasche commented on parsing/location.ml:
```diff
@@ -258,29 +263,46 @@ let print_loc ppf loc =
+    setup_colors ();
     print ppf loc;
-    fprintf ppf "Warning %a@." Warnings.print w
+    pp_open_tag ppf "warning";
```
is there a particular reason for not using a format string here?

xavierleroy commented:
For what it's worth, keep in mind that the Windows terminal does not support ANSI escapes, despite MSDOS's famous ANSI.SYS extension that made them ubiquitous:
http://stackoverflow.com/questions/16755142/how-to-make-win32-console-recognize-ansi-vt100-escape-sequences
So, one way or another, make sure that color support is turned off in a Windows terminal.

gasche commented on parsing/location.ml:
```diff
@@ -258,29 +263,46 @@ let print_loc ppf loc =
+
+let print_error_prefix ppf () =
+  setup_colors ();
+  pp_open_tag ppf "error";
```
Is there a particular reason for not using a format here? (I don't mind either the format or plain-`pp` styles, but mixing both is a bit surprising)

c-cube commented:
@gasche I'm moving to format strings, right.
@xavierleroy I don't have access to a windows machine, sadly. Do you know whether the TERM environment variable is set on windows? If it's not set (or equal to "dumb"), then colors will be disabled by default.

gasche commented on parsing/syntaxerr.ml:
```diff
@@ -28,32 +28,35 @@ let prepare_error = function
-          Location.error ~loc:opening_loc
-            (Printf.sprintf "Error: This '%s' might be unmatched" opening)
+          Location.errorf ~loc:opening_loc
+            "%a This '%s' might be unmatched" Location.print_error_prefix () opening
```
Could we export a `Location.errorf_prefixed` to factor these `Location.print_error_prefix`? There are a couple of calls to `Location.errorf` in the compiler without the `Error:` prefix (I see one in `ctype.ml` and three in `typetexp.ml`), and you should feel free to convert all of them to `errorf_prefixed` (it is inconsistent that they don't use `Error:`). Users of `raise_errorf` is a slightly different matter, it's probably best to leave them unchanged.

gasche commented:
@alainfrisch , any opinion on the error-printing code?

gasche commented:
(It may look like I have new remarks to make each time I look at the patch, but I'm actually happy with the convergence and thinking of adding it to my merge list.)

c-cube commented:
I think the `errorf_prefixed` is indeed a good idea. Glad you consider merging the patch! :)

dbuenzli commented:
> Do you know whether the TERM environment variable is set on windows? If it's not set (or equal to "dumb"), then colors will be disabled by default.

On a windows 10 tech preview IE virtual machine (available [here](https://www.modern.ie/fr-fr/virtualization-tools)). It seems that in `cmd.exe` `TERM` is not set at all, i.e. `Sys.getenv` raises `Not_found`.

gasche commented:
Onto the merge list!

We'll leave it to the future brave soul to provide support for whatever colors-in-terminal API Windows support - I know clang prints colored diagnostics on Windows as well.

gasche commented on driver/main_args.ml:
```diff
@@ -368,6 +368,18 @@ let mk_warn_help f =
+
+let mk_color_auto f =
+  "-color-auto", Arg.Unit f,
+  " Automatic mode for ANSI color codes for warning/error messages"
```
You should say that this is the default.

xavierleroy commented:
Re: Windows, a quick test shows that Cygwin shell windows have TERM=xterm and support ANSI escape sequences.  That's good!

alainfrisch commented:
> @alainfrisch , any opinion on the error-printing code?

Nope.

c-cube commented:
vim doesn't automatically change `$TERM` when it invokes the compiler (with `:make`), so it requires a workaround `:set makeprg=TERM=dumb\ make`. This is annoying.

The real solution would be to disable colors when stderr is not a tty, as mentioned above, but that requires `Unix.isatty`. Should I rather add a primitive in the compiler?

gasche commented:
I think having a primitive in the compiler would be nice -- it could go in `byterun/sys.c`. (For example, this would scale to adding Windows-specific logic in the future, whereas we wouldn't extend `Unix` with non-portable code.)

c-cube commented:
So you'd be ok with adding a primitive dedicated to, basically, "detecting whether some FD is color-ready"? Or whether a FD is an interactive terminal? That would currently reduce to calling "isatty" on unix, I think.

dbuenzli commented:
> - it could go in byterun/sys.c

If you do so please expose it in `Sys` this can benefit many other programs.

gasche commented:
Note that there is already code in the distribution to query terminal information, namely [utils/terminfo.ml](https://github.com/ocaml/ocaml/blob/trunk/utils/terminfo.ml) and the supporting C code [byterun/terminfo.c](https://github.com/ocaml/ocaml/blob/trunk/byterun/terminfo.c). I have no clue [how it is used](https://github.com/ocaml/ocaml/blob/trunk/parsing/location.ml#L91-L125) to nicely print locations, but it may be possible to extend the features of the module (which are currently just "does it support standout mode") to also giving information about color support, for example by adding a boolean parameter (named as a record field?) to the [Terminfo.Good_term](https://github.com/ocaml/ocaml/blob/trunk/utils/terminfo.ml#L18) case.

(`man terminfo` tells me that the variables to query are `setaf` (ability to set the foreground using an ASCI escape code) and `setab` (... background ..))

gasche commented:
P.S.: I was told ("dans l'oreillette") that some OCaml developers have white terminal background by default, so at some point you should check that your color choices work well with white backgrounds as well (I have my doubts about yellow). In writing the message above I wondered whether this could be queried, but that seems brittle.

c-cube commented:
**Better** support for white background can be added later, possibly by allowing the user to choose a set of colors among several ones (or define a CSS file for the compiler, why not!).

I'm going to take a look at terminfo.

ghost commented:
> I have no clue how it is used to nicely print locations

It does the same as the color printing code: it prints escape sequences to turn underline mode on/off. The only difference is that it uses the terminfo functions to do it

c-cube commented:
It is very strange that `Terminfo.setup` takes a channel as argument, but does not use it for anything (indeed, curses functions — I'm starting to understand the name of this library — do not take a filedescriptor as argument). Should I worry?

ghost commented:
@c-cube it uses it, in `terminfo_putc` which is called by `tputs`

c-cube commented:
Oh I see, it sets a global variable. Allow me to throw up a little. Anyway, I tried adding a "colors" capability to `Terminfo` (it seems to work), but this is still not sufficient:

``` sh
ocamlc foo.ml 2> some_file
```

should disable colors (they are sent to a file, not on the color-able terminal). So I think the conjunction of `Terminfo` and `isatty` is required.

ghost commented:
Yeah, the terminfo functions just look in the database.

Also it might be cleaner to use the same method for underline and colors, i.e. either use hardcoded escape sequences for both or use terminfo functions for both.

c-cube commented:
I'm having troubles compiling: if I put a C primitive in `sys.c`, it will not be linked in compiler-libs:

```
boot/ocamlrun boot/ocamlc -nostdlib -I boot  -compat-32 -o ocamlc \
   compilerlibs/ocamlcommon.cma compilerlibs/ocamlbytecomp.cma driver/main.cmo
File "_none_", line 1:
Error: Error while linking compilerlibs/ocamlcommon.cma(Misc):
The external function `caml_sys_is_interactive' is not available
```

On the other hand, I would really like this function to be part of the runtime, so that it could be exported in `Sys` (as asked by @dbuenzli). Ideas?

c-cube commented:
I could not fix the issues. Could someone that knows how OCaml's compilation process works please tell me why compilerlibs tries to link `caml_sys_is_interactive` but not, say, `caml_sys_file_exists`? grep tells me both names occur in exactly the same set of files…

ghost commented:
@c-cube, you can do this:
1. comment the external in `sys.mlp` (replace it with a dummy implem)
2. `make world`
3. `make bootstrap`
4. uncomment the external in `sys.mlp`
5. `make world`

Don't know if there is a better way...

c-cube commented:
It worked! Thanks a lot @diml . I had to upgrade the bootstrap so it includes the new primitive, `caml_sys_is_interactive` (basically a portable isatty).

c-cube commented:
Now, color printing is disabled if stderr is a file, a shell pipe, if TERM is "" or "dumb"...

dbuenzli commented:
> caml_sys_is_interactive

Just to say, this is potentially confusing with the existing boolean reference `Sys.interactive`.

c-cube commented:
It is exposed as `Sys.chan_is_interactive`, but I see your point. What name would be better?

c-cube commented:
side note: I have an itch to add colors to the testsuite, too ;)

dbuenzli commented:
> Sys.chan_is_interactive

`Sys.out_is_interactive : out_channel -> bool` would be more consistant with `Pervasives`' terminology.

Besides don't we also want this for input channels ? 

`Sys.{in,out}_channel_is_tty` ?

c-cube commented:
I don't know, does it make sense for an `in_channel` to refer to a terminal?

ghost commented:
Usually `stdin` refers to a terminal

dbuenzli commented:
> I don't know, does it make sense for an in_channel to refer to a terminal?

I think you may want to bufferize input differently (think e.g. about things like proposing completion in toplevels).

c-cube commented:
Right. Both can map to the same primitive anyway.

gasche commented:
I think exposing the new primitive in the stdlib is a bad idea. External users that want `isatty` can already use `Unix.isatty`, which does the job and has a good API. You should have `external chan_is_interactive` in `utils/misc.ml` where we need it, and leave the poor stdlib alone.

(It's not clear to me whether you also improved the portability of `isatty` by using a Windows equivalent. If this can be tested to actually work, we could add it to the Unix code.)

gasche commented on parsing/location.ml:
```diff
@@ -376,11 +409,13 @@ let () =
       | Sys_error msg ->
-          Some (errorf ~loc:(in_file !input_name) "Error: I/O error: %s" msg)
+          Some (errorf ~loc:(in_file !input_name)
+                  "%a I/O error: %s" print_error_prefix () msg)
```
`errorf_prefixed`?

gasche commented on parsing/location.ml:
```diff
@@ -376,11 +409,13 @@ let () =
             (errorf ~loc:(in_file !input_name)
-               "Error: Some fatal warnings were triggered (%d occurrences)" n)
+               "%a Some fatal warnings were triggered (%d occurrences)"
+               print_error_prefix () n)
```
`errorf_prefixed`?

dbuenzli commented:
> I think exposing the new primitive in the stdlib is a bad idea.

It can be useful in more exotic OS environments where the full `Unix` library is not available but reimplementing the primitive can make sense.

c-cube commented:
It would be nice to have the opinion of more maintainers over this primitive.

gasche commented:
In preparation of an eventual merge, it would be nice if you could eventually rebase the patch series in a few independent patches with a good commit message. We could have, for example, one patch for `Colors`, one patch for the actual message printing (`location` and its users), one patch for the command-line stuff (and ocamlbuild and OCAMLPARAM), and one patch for the isatty stuff.

xavierleroy commented:
Some random comments.  
- Don't use terminfo.  Up to the late 80's, there were a lot of (physical) terminals (of the kind that's attached via a RS232 serial link), each with its own quirky command language.  The termcap and later terminfo databases were essential for Unix programs to be compatible with as many terminals as possible.  Nowadays, our terminals are just windows in a GUI, and they all talk ANSI escape codes (or are dumb and talk no codes at all, like Microsoft's command window).  So, go ahead, send ANSI escapes if TERM=ansi or TERM=xterm.  While you're at it, you could rewrite Location.highlight_locations in the same style.  This way I could remove this terminfo historical junk from OCaml altogether.
- I agree with gasche that Unix.isatty is fine for other programs, and that whatever primitive you introduce for isatty() should not be exported in the stdlib, at least initially.  Keep it simple, just wrap "isatty(0)" as a unit -> bool OCaml function for compiler internal use only.
- By the was, "isatty" is a time-honored name that, for Unix people, is clearer than "chan_is_interactive" or most other names you could come up with.
- +1 for the remark about white backgrounds.  Any color choice that is not readable on a white background will be rejected.

dbuenzli commented:
> Keep it simple, just wrap "isatty(0)" as a unit -> bool OCaml function for compiler internal use only.

You want to do that on stderr instead, so that if it is redirected to a file you don't produce ANSI escapes.

c-cube committed:
update changelog for PR#207 (coloring messages)

c-cube committed:
add primitive caml_sys_is_interactive; fix bootstrap

c-cube committed:
add Color module Misc

use bold magenta for warnings (works on white background)

c-cube committed:
colorize error and warning messages, refactor Location

c-cube committed:
add option handling for colors in compiler, OCAMLPARAM and ocamlbuild

use -color, -no-color, -color-auto in the compiler
use "0", "1" or "a" for the "color" option in OCAMLPARAM
add tags 'color', 'color_auto', 'no_color' to ocamlbuild

c-cube commented:
Here we are, I did a big interactive rebase. There is a new primitive in `sys.c` but it is not exported in `Sys`. The color for warnings is bold magenta, which I hope will be visible on all terminals (it is similar to what gcc and clang do).

About `Terminfo`, I removed the color handling part; however, I don't know how totally removing `Terminfo` from the compiler would work. @xavierleroy do you have in mind using ansi codes for moving the cursor in the terminal?

gasche commented on byterun/sys.c:
```diff
@@ -485,3 +488,20 @@ CAMLprim value caml_sys_read_directory(value path)
+
+/* Return true if the value is a filedescriptor (int) that is
+ * (presumably) open on an interactive terminal */
+CAMLprim value caml_sys_is_interactive(value chan)
```
I think `caml_sys_isatty` would be a better name, but that is nitpicking.

ghost commented:
> About `Terminfo`, I removed the color handling part; however, I don't know how totally removing `Terminfo` from the compiler would work. @xavierleroy do you have in mind using ansi codes for moving the cursor in the terminal?

That's already what terminfo.c does, only it does it using tputs. If we use ansi codes directly for colors we might as well use ansi codes directly for underline and cursor movements.

xavierleroy commented:
Yes, as @diml wrote, I was thinking of hard-coding ANSI escape sequences for moving the cursor and going into/out of standout mode.  One difficulty I didn't anticipate is getting the number of lines of the terminal.  Terminfo has some special magic to query the current size of xterm-like terminal windows.  Perhaps we can hard-wire a conservative 24 lines...  but it's less satisfactory.

ghost commented:
For the terminal size we can use the `TIOCGWINSZ` ioctl (or `GetConsoleScreenBufferInfo` on windows), as done here: https://github.com/diml/lambda-term/blob/master/src/lTerm_term_stubs.c

edwintorok commented:
@gasche "On the other hand, checking how other projects handle this question (eg. clang?) may suffice to have a good/consensual answer." Guess [I am to blame for the color diagnostics in clang](http://lists.cs.uiuc.edu/pipermail/llvm-commits/Week-of-Mon-20090525/078183.html). It was a long time ago, but IIRC the motivation was that clang started to display source code fragments in error messages, and it became hard for me to follow what the error message is and which file it come from, and using colors/bold for the file and error message made it easier to read.

When I have [initially added color diagnostics to LLVM/Clang](https://github.com/llvm-mirror/llvm/commit/e8ebb0fe1bba0fdff7475d98e1f8a04804b0b056) I used the simple approach with checking `isatty` and that `TERM` is not `dumb`. Several years later someone changed it to [use ncurses for the detection](https://github.com/llvm-mirror/llvm/commit/f7364d5833f700eefd7919326bfa5724da1461c3) and [now it uses `tigetnum("colors")`](https://github.com/llvm-mirror/llvm/commit/8d8bdff6d7eccb05bf16e18141263ee72ea8296b).  Another way would be to call `tput colors`. Unfortunately the commit message doesn't say whether the `TERM` based detection was a problem in practice or not.
FWIW `git` and `grep` still use the `isatty` and `TERM` based test, I think it should be fine for OCaml too.

Also see [how colors were handled in Windows without ANSI escapes](https://github.com/llvm-mirror/llvm/commit/e8ebb0fe1bba0fdff7475d98e1f8a04804b0b056#diff-27e93856cd53aadc72c126230288e431R149). I no longer have/use Windows though so I don't know if that still works on latest versions and I doubt it'd be worth the trouble to add windows-specific color support to OCaml.

c-cube commented:
I would like to keep this PR relatively simple, and I believe it is ready — as its name indicates it only adds the bare minimum for color handling. I think adding a second primitive to check the number of lines and get rid of terminfo is a good idea, but it needs not be done right now; same goes for allowing the user to choose colors or other things.

gasche commented:
Yeah, I also think the current state is fine.

@edwintorok thanks for the write-up!

gasche commented:
Fun fact: because your rebased the bootstrap that was done on the old patch state, applying this patch series makes the compiler print warning in yellow, while the code specifies purple -- with no mention of yellow in the code or its history. On trusting trust, tastes and colors.

gasche commented:
Sorry to nitpick, but the current patch series isn't really usable as is -- for example, [this commit](https://github.com/c-cube/ocaml/commit/ec34fc9cc80f1765c962054615c4598afcce1abb) is not at all what it says. Would it be possible to have another rebase? I will do it myself otherwise, but probably not today.

Other points I noticed when trying to rebase myself this morning:
- the code says `caml_sys_is_interactive`, the commit message says `caml_sys_isatty`, you should pick one
- I think there is a useless `let color = ref false` definition in one commit

gasche commented:
Also, I noticed during my testing that, when printing a location, the characters are not highlighted with the rest of the location (this corresponds to a small bug in `print_loc`).

c-cube commented:
I will do the rebase and change the bad commit/commit messages. What do you mean by non-highlighted 'characters'?

gasche commented:
Only the line numbers are highlighted, not the character range.

c-cube committed:
fix small bug in `print_loc`

c-cube commented:
Right, fixed. About the bootstrap, I admit it's funny to have yellow colors (it illustrates the idea of the compiler hidden backdoor perfectly imho); I can change the bootstrap again if you want, either using the latest version or the one just after `caml_sys_is_interactive` was added.

gasche commented:
I think it's nice to bootstrap with the color-enabled compiler, as it ensures that other maintainers will see the effect of the change -- early adopters! But then you should use the right colors. (Bootstrapping before is fine as well.)

You should also merge "add option handling" and "add color handling in OCAMLPARAM and ocamlbuild". Thanks for the changelog message. Would it be possible to also have a test in the testsuite that checks that the behavior is as expected for `-color` and `-no-color`?

(I'm thinking of adding two changes on top of this PR in the future: change the option handling to use `-color {always,never,auto}` instead, and add a color for hints. I might also experiment with having locations underlined instead of bold, or maybe not emphasized at all, because right now they are a bit invasive visually.)

c-cube commented:
Ok, will do. About the bold locations, I followed what clang and gcc do, and I don't find it too invasive. It will be easy to change later anyway.

Is there something that would help me write tests that set specific options to the compiler before compiling the sample?

gasche commented:
Now that I think of it, testsuite would be a pain because you'd have to disable them under Windows and other unsupported environment. I think it's not worth it, so feel free to skip these. (If color get broken, people will complain anyway, hopefully...).

An example of test controlling the command-line parameters of the compiler is [lib-dynlink-bytecode](https://github.com/ocaml/ocaml/blob/trunk/testsuite/tests/lib-dynlink-bytecode/Makefile).

c-cube commented:
Ok. Then I think  the current set of commits is ready (I re-bootstraped and fixed the location bug, please check!).

gasche commented:
Merged in trunk. Thanks for the patch!

c-cube commented:
apologies for the bad rebasing (I did rebase a lot tbh, it must have gone wrong at some iteration).

gasche commented:
Based on [this caml-list discussion](https://sympa.inria.fr/sympa/arc/caml-list/2015-09/msg00164.html), it seems that the current patch breaks the ability to define a custom margin for toplevel messages. We should fix that, and add a regression test to the testsuite.

c-cube commented:
I don't think it _breaks_ this feature, I think it was already broken. Formatting was done using `ksprintf` and buffers. I can try to fix it nonetheless.

gasche commented:
Indeed, it seems you were right and the change happened between 4.01 and 4.02. Sorry for jumping to the wrong conclusions.
